src/gc/gc.d

```diff
@@ -3130,21 +3068,122 @@ struct Pool
+        info.attr = getBits(cast(size_t)(offset >> shiftBy));
+
+        return info;
+    }
+
+    void runFinalizers(in void[] segment) nothrow
+    {
+        foreach (pn; 0 .. npages)
         {
-            size = bPageOffsets[pagenum] * PAGESIZE;
+            Bins bin = cast(Bins)pagetable[pn];
+            if (bin >= B_PAGE)
+                continue;
+
+            immutable size = binsize[bin];
+            auto p = baseAddr + pn * PAGESIZE;
+            const ptop = p + PAGESIZE;
+            auto biti = pn * (PAGESIZE/16);
+            immutable bitstride = size / 16;
+
+            GCBits.wordtype toClear;
+            size_t clearStart = biti >> GCBits.BITS_SHIFT;
+            size_t clearIndex;
+
+            for (; p < ptop; p += size, biti += bitstride, clearIndex += bitstride)
+            {
+                if (clearIndex > GCBits.BITS_PER_WORD - 1)
+                {
+                    if (toClear)
+                    {
+                        clrBitsSmallSweep(clearStart, toClear);
+                        toClear = 0;
+                    }
+
+                    clearStart = biti >> GCBits.BITS_SHIFT;
+                    clearIndex = biti & GCBits.BITS_MASK;
+                }
+
+                if (!finals.test(biti))
+                    continue;
+
+                auto q = sentinel_add(p);
+                uint attr = getBits(biti);
+
+                if(!rt_hasFinalizerInSegment(q, size, attr, segment))
+                    continue;
+
+                rt_finalizeFromGC(q, size, attr);
+
+                toClear |= GCBits.BITS_1 << clearIndex;
+
+                debug(COLLECT_PRINTF) printf("\tcollecting %p\n", p);
+                //log_free(sentinel_add(p));
+
+                debug (MEMSTOMP) memset(p, 0xF3, size);
+                freebits.set(biti);
+            }
+
+            if (toClear)
+            {
+                clrBitsSmallSweep(clearStart, toClear);
+            }
         }
-        return size;
     }
 
-    @property bool isFree() const pure nothrow
+    /**
+    * Allocate a page of bin's.
+    * Returns:
+    *           head of a single linked list of new entries
+    */
+    List* allocPage(Bins bin) nothrow
     {
-        return npages == freepages;
+        size_t pn;
+        for (pn = searchStart; pn < npages; pn++)
+            if (pagetable[pn] == B_FREE)
+                goto L1;
+
+        return null;
+
+    L1:
+        searchStart = pn + 1;
+        pagetable[pn] = cast(ubyte)bin;
+        freepages--;
+
+        // Convert page to free list
+        size_t size = binsize[bin];
+        void* p = baseAddr + pn * PAGESIZE;
+        void* ptop = p + PAGESIZE - size;
+        auto first = cast(List*) p;
+
+        for (; p < ptop; p += size)
+        {
+            (cast(List *)p).next = cast(List *)(p + size);
+            (cast(List *)p).pool = &base;
+        }
+        (cast(List *)p).next = null;
+        (cast(List *)p).pool = &base;
+        return first;
```

Comment:
Ah OK, took me a second look to see, that this chains the list in forward order.